design-proposals/2022-10-27-rendering-modularization.md

```diff
@@ -0,0 +1,508 @@
+- Replace existing shaders without having to recompile the whole set.
+- Let developers communicate with their own shaders.
+
+#### For GLES we’ll need to:
+- Push aside the existing Program hierarchy and rename it with a GLES extension.
+- Allow for named uniforms for new Programs<br>We can probably ignore that requirement for the existing shaders and just wrap them
+
+### Shader(Program) Registry
+
+_Addresses core functionality [#1](#core) and [#2](#core)._
+
+Once developers can create their own shaders, we’ll need a way of managing them, thus the Shader Registry.  This registry will associate a given shader (Program) with a name.  Basic functionality will include:
+- Add a new shader by name
+- Replace an existing shader by name
+- Query a shader by name
+- Publish the well known names of the core shaders for the basic functionality
+
+Layers already go through a level of indirection to instantiate or reference their associated programs.  The shader registry could be inserted into that get*LayerPrograms() pattern fairly easily.
+
+![Shader Registry](resources/fig1.png)
+
+The GLES version of the Shader Registry can encapsulate the simple (but fairly weird) logic of shader source lookup that exists now.
+
+#### The way it is now
+
+The layers more or less pull their shaders out of a compressed block of text.  Clever, but difficult to access.
+
+#### Required Changes
+
+We'll need to make a Shader Registry class, of course, but that's fairly simple.  It should be thread safe, but you're mostly going to be using it on the main thread.
+
+The layers should pull their shaders out of the shader on demand.  This will give developers a chance to replace or add shaders before they're needed.
+
+### Rendering Passes
+
+_Addresses core functionality [#5](#core)._
+
+Modern graphics pipelines use more than one pass to create a visual representation.  It's actually pretty unusual to have a single rendering pass in your toolkit.  These are used for a dizzying array of purposes, including lightning effects, shadows, atmosphere and many others.  We (Wet Dog Weather) use separate rendering passes to draw data intermingled with map elements.
+
+Without getting too deep into specifics, early rendering passes allow the developer to use the power of the rasterizer for their own data.  Later rendering passes are typically used to decorate the map with effects.
+
+#### The way it is now:
+The [low level rendering logic](https://github.com/maplibre/maplibre-gl-native/blob/main/src/mbgl/renderer/renderer_impl.cpp) in the toolkit seems to only support one explicit rendering pass.  
+
+However, there is the [RenderPass](https://github.com/maplibre/maplibre-gl-native/blob/main/src/mbgl/gl/render_pass.hpp) so some notion of this exists in the toolkit, but it's not clear how complete that is.  It may depend on ordering to work things out, rather than the explicit command buffer filling and fences we would use in a more modern approach.
```

Comment:
I guess that RenderPass referes a different concept compared to Vulkan/Metal's rendering passes.

E.g. even if the tookit uses right now multiple render passes, we probably still only need a single one.
So we should check that toolkit's RenderPass = Metal RenderPasses

Reply:
Agree, a Metal RenderPass can finish a frame draw. But some features need an extra pass, the heatmap layer and the FillExtrusionLayer need some additional render passes to draw some offscreen texture, but the other layers only draw on the map view. In other words, mbgl::RenderPass::Pass3D is some metal render pass for offscreen textures, but the others mbgl::RenderPass::Opaque and  mbgl::RenderPass::Translucent work in one metal render pass.

Reply:
The naming is confusing.  It's sort of a render pass, but not as an explicit one as you'd normally have in these sorts of toolkits.

For example, the heatmap layer does do that, but the main rendering "loop" isn't as aware of that as it should be.  And it's hard for other parts of the system to make use of that same functionality.  Ideally you should be able to designate features as being drawn in specific passes and then used as inputs to other passes.

The difference is that your heatmap logic has to set that up and be aware of it.  Rather, we'd prefer that tiled image layers be fairly generic and drawable to whatever target you designate.  Then heatmap (and other) setup moves up a level and we make it possible to do that sort of thing with other data sets outside of the core rendering logic.